WHL scale testing - Crash dumps

chris-raitano opened the pull request:
## Summary
<!--- Provide a detailed description of a change. -->
<!--- What was the problem? -->
<!--- Why are we fixing the problem? -->
<!--- How was the problem solved? -->

This PR creates a container for generating crash dumps for scale testing. 

This builds off of PR #1066 _WHL Scale Testing - Log GenerationInfra_. Reviewers should view the [diff](https://github.com/microsoft/Docker-Provider/compare/craitano/log-generation-infra...craitano/crash-dump-generation) between these branches to see the changes made by this PR 

### Type of change:
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Feature work
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation
- [ ] Developer Experience
- [x] Test
- [ ] Logging/Metrics
- [x] Automation 

<!--- ## Revisions (Optional) -->
<!--- Use this optional section for highlighting changes between PR revisions which require further detail. -->
<!--- For example, does the next revision contain changes not related to PR comments which reviewers should be aware of? -->

## Developer Notes
<!--- Are there technical details or context reviewers should be aware of? -->
<!--- For example, were some choices made due to framework constraints? -->
<!--- Is a new dependency being introduced? -->
<!--- Is this PR dependent on another change? -->

## Testing
<!--- How was this change validated before PR -->
- [ ] Unit tests
- [ ] Integration tests
- [x] Manual tests (add details)

Follow the instructions in Section 5 of [whl-scale-test.md](https://github.com/microsoft/Docker-Provider/blob/craitano/crash-dump-generation/test/whl-scale-tests/whl-scale-test.md#5-generating-logs-for-testing) to deploy the crash dumps generator.

cacarlt committed:
Draft of infra setup

cacarlt committed:
Updating branch with doc and cleanup script

cacarlt committed:
Scripting Geneva File Updates

cacarlt committed:
Update Geneva Template for Crash Dumps

cacarlt committed:
Finishing Geneva Scenarios

cacarlt committed:
Updating deploy-infra script

cacarlt committed:
Adding new scale test suite script

cacarlt committed:
Reducing duplicate code

cacarlt committed:
Scale Test Suite Script Works

cacarlt committed:
Updating WHL Scale Test doc

cacarlt committed:
Quick fix

cacarlt committed:
Updating Doc

cacarlt committed:
Another md update

cacarlt committed:
Update prepare script

cacarlt committed:
Finished Section 2

cacarlt committed:
Update from Feedback

cacarlt committed:
Updating Section 3 and password gen

cacarlt committed:
Update Crash Dump File

cacarlt committed:
Adding XML Template for ETW and Event Logs

cacarlt committed:
Clean Infra Updated

cacarlt committed:
Updating usage of password and common sciprt

chris-raitano committed:
Log generation infrastructure

chris-raitano committed:
Add ACR parameter

chris-raitano committed:
Add ACR login and gitignore temp folder

chris-raitano committed:
Set up crash dumps scale testing

cacarlt committed:
Updating deploy test suite script

cacarlt committed:
Updated troubleshooting and validation steps

cacarlt committed:
PR Update

cacarlt committed:
Updating Troubleshooting and Validation images

chris-raitano committed:
Merge branch 'cacarlt/scale-infra' into craitano/log-generation-infra

cacarlt committed:
Updates related to PR comments

chris-raitano committed:
Usability improvements

cacarlt commented on test/whl-scale-tests/whl-scale-test.md:
Make sure to add new script and config maps usage to the markdown file.

alrodrig1 commented on test/whl-scale-tests/deploy-infra.ps1:
```diff
@@ -0,0 +1,111 @@
+
+# Required for Windows node pool and could be used to troubleshoot any issues
+Write-Host "Creating random password for Host login"
+$password = Get-RandomPassword 16 1 1 1 1
```
If you want to avoid having to generate passwords on the fly every time you run this, you could just generate a strong password (manually, outside this script), put it in Key Vault, and fetch it here.

cacarlt replied:
We have found this is the best user experience as this is cleaned up with the removal of the resource group and if the password is less likely to fail the deployment of the aks cluster.

cacarlt committed:
Using Windows 2019 and 2022

cacarlt commented on test/whl-scale-tests/crash-dumps/deploy.ps1:
```diff
@@ -0,0 +1,50 @@
+  Write-Host "End:Downloading Nuget Package: $PACKAGE_NAME"
+}
+
+function Cleanup{
+  # We can leave items in the temp folder since those will be cleaned up by the infrastructure. But should clean up anything else
+  Remove-Item $TmpZipFile -Force
+}
```
What files is needing to be cleaned up? I am only seeing the .zip

chris-raitano replied:
Everything else is in Temp folder. Moved the zip file there too

cacarlt commented on test/whl-scale-tests/common.ps1:
```diff
@@ -0,0 +1,170 @@
+}
+
+<#
+.SYNOPSIS
```
Not sure we need a synopsis and a description

cacarlt commented on test/whl-scale-tests/crash-dumps/deploy.ps1:
```diff
@@ -0,0 +1,50 @@
+$DownloadName = "$PACKAGE_NAME.$PACKAGE_VERSION" # Expected folder/file name of the downloaded nuget package
+$TmpDir = Get-TempDir
+$DownloadPath = Join-Path $TmpDir "$DownloadName\$DownloadName.nupkg"
+$TmpZipFile="$PSScriptRoot\crashdumpgenerator.zip" # Can't go in temp directory since dockerfile won't have access
```
If we pass in a different context we should be able to see the TempDir. Also since you are creating the temp dir you can put it where it makes the most sense to improve the usage of the automation.

cacarlt commented on test/whl-scale-tests/crash-dumps/main.ps1:
```diff
@@ -0,0 +1,49 @@
+.DESCRIPTION
+Uses Azure watson's hugedump exe to generate crash dumps. Requires the following Env vars are set:
+* CRASHD_NUM_CRASHES - Number of crash dumps to generate
+* CRASHD_ALLOC_STRATEGY - Strategy to use for memory allocation. Options are:
+      va : allocate using VirtualAlloc
+      ha : allocate using HeapAlloc
+      sa : allocate on threads' stacks
```
This clarification belongs in the configmap as well

cacarlt replied:
After talking with Chris offline we have decided to document this inside the md file for clarity.

chris-raitano committed:
Merge branch 'craitano/log-generation-infra' into craitano/crash-dump-generation

cacarlt commented on test/whl-scale-tests/crash-dumps/main.ps1:
```diff
@@ -0,0 +1,49 @@
+Write-Host "$(Get-Timestamp) Crash dump generation completed"
+
+# Keep pod alive to prevent crash loop
+while($true){
+  start-sleep ([uint16]::MaxValue)
+}
```
Do we need the container alive once it is done creating crash dumps? If so we shouldn't have an infinite loop as this isn't the best usage of our restricted resource environment. We should instead use notepad.exe as we do in WHL.

cacarlt replied:
After talking with Chris offline we have landed on using the notepad.exe as CI does with all their long living containers.

cacarlt commented on test/whl-scale-tests/deploy-log-generators.ps1:
```diff
@@ -0,0 +1,63 @@
+
+if($CrashDumps -or $all){
+  Write-Host "START:Deploying Crash Dump Generator"
+  & "$PSScriptRoot\crash-dumps\deploy.ps1" "$ACR/generatecrashdumps:latest" ltsc2022 whl-crashd crashd
```
image.png

chris-raitano committed:
Update to use merged infra changes

cacarlt committed:
Removed clean-infra optional parameter

cacarlt committed:
Limiting password gen

cacarlt commented on test/whl-scale-tests/deploy-log-generators.ps1:
```diff
@@ -0,0 +1,124 @@
+  
+}
+
+az login
```
Add the set subscription to the script after this line

chris-raitano replied:
It's not needed for this script. ACR names are unique across all subscriptions/resource groups so the acr login works regardless of the current subscription

cacarlt commented on test/whl-scale-tests/deploy-log-generators.ps1:
```diff
@@ -0,0 +1,124 @@
+}
+
+az login
+az acr login -n $acrName
```
Open up docker before running this as it will fail without it.

cacarlt committed:
Typo changes

alrodrig1 commented on test/whl-scale-tests/crash-dumps/main.ps1:
```diff
@@ -0,0 +1,49 @@
+  [string]::IsNullOrEmpty($numCrashes) -or 
+  [string]::IsNullOrEmpty($strategy) -or 
+  [string]::IsNullOrEmpty($totalSize ) -or 
+  [string]::IsNullOrEmpty($numCrashes)
```
should this be `$portionSize` instead of `$numCrashes`?

chris-raitano committed:
Merge branch 'cacarlt/scale-infra' into craitano/log-generation-infra

chris-raitano committed:
Add documentation and address PR comments

chris-raitano committed:
Remove duplicate variables that are set in common.ps1

chris-raitano committed:
Merge branch 'craitano/log-generation-infra' into craitano/crash-dump-generation

chris-raitano committed:
Addressing PR comments

chris-raitano committed:
Add Start-Docker function and clean up function headers

chris-raitano committed:
Merge branch 'craitano/log-generation-infra' into craitano/crash-dump-generation

cacarlt commented on test/whl-scale-tests/deploy-log-generators.ps1:
```diff
@@ -0,0 +1,120 @@
+.EXAMPLE
+Deploy-LogGenerator exampleacr.azurecr.io/generatelogs:latest LogGenerator log-generation whl-logs
+#>
+function buildAndDeploy {
```
Is it best practice to use camel case with local functions?

chris-raitano replied:
I think Pascal case is best practice. Updated

cacarlt commented on test/whl-scale-tests/whl-scale-test.md:
```diff
@@ -0,0 +1,302 @@
+# Windows Host Logs Scale Test
+
+## Prerequisites
```
If the script wont install/download nuget.exe for the user it should be listed here as Prereq

cacarlt commented:
Not sure what is busted but I am unable to get the crash dump image

Login Succeeded
START:Deploying Crash Dump Generator
Applying settings config map to whl-crashd
configmap/whl-scale-test-settings unchanged
Daemonset whl-crash-dump-generator already exists. Restarting with new configuration.
daemonset.apps/whl-crash-dump-generator restarted
END:Deploying Crash Dump Generator
START:Deploying ETW Generator
END:Deploying ETW Generator
START:Deploying Event Log Generator
END:Deploying Event Log Generator
START:Deploying Text Log Generator
END:Deploying Text Log Generator
PS C:\Git\Docker-Provider\test\whl-scale-tests> kubectl describe pod -n whl-crashd whl-crash-dump-generator-qwslz 
Error from server (NotFound): pods "whl-crash-dump-generator-qwslz" not found
PS C:\Git\Docker-Provider\test\whl-scale-tests> kubectl get pods -n whl-crashd                                    
NAME                             READY   STATUS             RESTARTS   AGE
host-logs-windows-gqhvg          1/1     Running            0          73m
whl-crash-dump-generator-w56qv   0/1     ImagePullBackOff   0          15s


Looks like the image never made to my acr
![image](https://github.com/microsoft/Docker-Provider/assets/69869281/2d10802a-59c1-4aca-813a-49658569d683)

I will leave my resources up for testing

UPDATE: No longer an issue

chris-raitano committed:
Add -ApplyConfigChanges flag for redeploying log generators

chris-raitano committed:
Merge branch 'craitano/log-generation-infra' into craitano/crash-dump-generation

chris-raitano commented:
> Not sure what is busted but I am unable to get the crash dump image
> 
> Login Succeeded START:Deploying Crash Dump Generator Applying settings config map to whl-crashd configmap/whl-scale-test-settings unchanged Daemonset whl-crash-dump-generator already exists. Restarting with new configuration. daemonset.apps/whl-crash-dump-generator restarted END:Deploying Crash Dump Generator START:Deploying ETW Generator END:Deploying ETW Generator START:Deploying Event Log Generator END:Deploying Event Log Generator START:Deploying Text Log Generator END:Deploying Text Log Generator PS C:\Git\Docker-Provider\test\whl-scale-tests> kubectl describe pod -n whl-crashd whl-crash-dump-generator-qwslz Error from server (NotFound): pods "whl-crash-dump-generator-qwslz" not found PS C:\Git\Docker-Provider\test\whl-scale-tests> kubectl get pods -n whl-crashd NAME READY STATUS RESTARTS AGE host-logs-windows-gqhvg 1/1 Running 0 73m whl-crash-dump-generator-w56qv 0/1 ImagePullBackOff 0 15s
> 
> Looks like the image never made to my acr ![image](https://user-images.githubusercontent.com/69869281/269803426-2d10802a-59c1-4aca-813a-49658569d683.png)
> 
> I will leave my resources up for testing

This could happen for a variety of reasons (acr login failed, docker build failed, etc). The problem is when you run deploy-log-generators.ps1 a 2nd time it skips the docker image build and just restarts the existing daemonset. This behavior assumes the build and deployment was successful the 1st time without an option to fix it when something failed

I've updated deploy-log-generators.ps1 to now rebuild the image and deploy a new daemonset rather than just restarting the existing daemonset. This way you can at least retry when it fails. I also added an -ApplyConfigChanges flag for when you do want to restart the daemonset without a full build

chris-raitano committed:
Merge branch 'sky-dev' into craitano/crash-dump-generation

cacarlt reviewed (approved):
Ready to sync with sky-dev and then merge.

cacarlt commented:
[celebrate]     Cameron Carlton reacted to your message:
________________________________
From: Chris Raitano ***@***.***>
Sent: Monday, October 9, 2023 9:22:19 PM
To: microsoft/Docker-Provider ***@***.***>
Cc: Cameron Carlton ***@***.***>; Comment ***@***.***>
Subject: Re: [microsoft/Docker-Provider] WHL scale testing - Crash dumps (PR #1067)


Merged #1067<https://github.com/microsoft/Docker-Provider/pull/1067> into sky-dev.

—
Reply to this email directly, view it on GitHub<https://github.com/microsoft/Docker-Provider/pull/1067#event-10596108670>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQVB5YM6VWWWCDEAESSUGZ3X6RTIXAVCNFSM6AAAAAA4ZAJVECVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJQGU4TMMJQHA3DOMA>.
You are receiving this because you commented.Message ID: ***@***.***>
